app/services/events/stores/aggregated_clickhouse/weighted_sum_query.rb

```diff
@@ -0,0 +1,210 @@
+            SELECT
+              toDateTime64(:from_datetime, 5, 'UTC') as timestamp,
+              toDecimal128(:initial_value, :decimal_scale) as difference
+          SQL
+        end
+
+        def end_of_period_value_sql
+          <<-SQL
+            SELECT
+              toDateTime64(:to_datetime, 5, 'UTC') as timestamp,
+              toDecimal128(0, :decimal_scale) as difference
+          SQL
+        end
+
+        def period_ratio_sql
+          <<-SQL
+            if(
+              -- NOTE: duration in seconds between current event and next one
+              date_diff('seconds', timestamp, leadInFrame(timestamp, 1, toDateTime64(:to_datetime, 5, 'UTC')) OVER (ORDER BY timestamp ASC ROWS BETWEEN CURRENT ROW AND UNBOUNDED FOLLOWING)) > 0,
+
+              -- NOTE: cumulative sum from previous events in the period
+              (SUM(difference) OVER (ORDER BY timestamp ASC ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW))
+              *
+              -- NOTE: duration in seconds between current event and next one - using end of period as final boundaries
+              date_diff('seconds', timestamp, leadInFrame(timestamp, 1, toDateTime64(:to_datetime, 5, 'UTC')) OVER (ORDER BY timestamp ASC ROWS BETWEEN CURRENT ROW AND UNBOUNDED FOLLOWING))
+              /
+              -- NOTE: full duration of the period
+              #{charges_duration.days.to_i},
+
+              -- NOTE: duration was null so usage is null
+              toDecimal128(0, :decimal_scale)
+            )
+          SQL
+        end
+
+        def grouped_events_cte_sql(initial_values)
+          <<-SQL
+            WITH events_data AS (
+              (#{grouped_initial_value_sql(initial_values)})
+              UNION ALL
+              (#{
+                events_sql(
+                  ordered: true,
+                  select: [
+                    Arel::Nodes::NamedFunction.new(
+                      "toJSONString",
+                      [arel_enriched_table[:sorted_grouped_by]]
+                    ).as("grouped_by"),
+                    arel_enriched_table[:timestamp].as("timestamp"),
+                    arel_enriched_table[:decimal_value].as("difference")
+                  ]
+                )
+              })
+              UNION ALL
+              (#{grouped_end_of_period_value_sql(initial_values)})
+            )
+          SQL
+        end
+
+        def grouped_initial_value_sql(initial_values)
+          values = initial_values.map do |initial_value|
+            [
+              "'#{formated_groupes_values(initial_value)}'",
+              "toDateTime64(:from_datetime, 5, 'UTC')",
+              "toDecimal128(#{initial_value[:value]}, :decimal_scale)"
+            ].join(", ")
+          end
+
+          tuple_select_sql(values)
+        end
+
+        def grouped_end_of_period_value_sql(initial_values)
+          values = initial_values.map do |initial_value|
+            [
+              "'#{formated_groupes_values(initial_value)}'",
```

Comment:
I feel like there's a risk of SQL injection here. Having a `'` in the group might break this query.